horovod/spark/common/util.py

```diff
@@ -0,0 +1,546 @@
+    if num_processes <= 0 or partitions_per_process <= 0:
+        raise ValueError('num_proc={} and partitions_per_process={} must both be > 0'
+                         .format(num_processes, partitions_per_process))
+
+    if not label_columns:
+        raise ValueError('Parameter label_columns cannot be None or empty')
+
+    num_partitions = num_processes * partitions_per_process
+    if verbose:
+        print('num_partitions={}'.format(num_partitions))
+
+    for col in label_columns:
+        if col not in df.columns:
+            raise ValueError('Label column {} does not exist in this DataFrame'.format(col))
+
+    if feature_columns is None:
+        feature_columns = [col for col in df.columns if col not in set(label_columns)]
+
+    key = _training_cache.create_key(df, store, validation)
+    with _training_cache.lock:
+        _training_cache.set_in_use(key, True)
+        if _training_cache.is_cached(key, store):
+            dataset_idx = _training_cache.get_dataset(key)
+            train_rows, val_rows, metadata, avg_row_size = _training_cache.get_dataset_properties(dataset_idx)
+            train_data_path = store.get_train_data_path(dataset_idx)
+            val_data_path = store.get_val_data_path(dataset_idx)
+            if verbose:
+                print('using cached dataframes for key: {}'.format(key))
+                print('train_data_path={}'.format(train_data_path))
+                print('train_rows={}'.format(train_rows))
+                print('val_data_path={}'.format(val_data_path))
+                print('val_rows={}'.format(val_rows))
+        else:
+            dataset_idx = _training_cache.next_dataset_index(key)
+            train_data_path = store.get_train_data_path(dataset_idx)
+            val_data_path = store.get_val_data_path(dataset_idx)
+            if verbose:
+                print('writing dataframes')
+                print('train_data_path={}'.format(train_data_path))
+                print('val_data_path={}'.format(val_data_path))
+
+            schema_cols = feature_columns + label_columns
+            if sample_weight_col:
+                schema_cols.append(sample_weight_col)
+            if isinstance(validation, str):
+                schema_cols.append(validation)
+            df = df[schema_cols]
+
+            metadata = _get_metadata(df)
+
+            to_petastorm = to_petastorm_fn(schema_cols, metadata)
+            train_df = df.rdd.map(to_petastorm).toDF()
+            train_df, val_df, validation_ratio = _train_val_split(train_df, metadata, validation)
+
+            train_partitions = max(int(num_partitions * (1.0 - validation_ratio)),
+                                   num_processes)
+            if verbose:
+                print('train_partitions={}'.format(train_partitions))
+
+            train_df \
+                .coalesce(train_partitions) \
+                .write \
+                .mode('overwrite') \
+                .parquet(train_data_path)
```

Comment:
what about saving failed?  The path should be created if failed. Should we wrap this with `try catch` and delete the failed path.

Reply:
I think it would be difficult to cleanup this path, because during writing Spark will write to a unique temporary file (if I remember correctly), then move it when writing completes.  I am not sure how to get the name of that temp file.

In most cases, it shouldn't be an issue because we will reuse the same paths for successive attempts.

Reply:
OK, that makes sense.